### DISCUSSION TITLE
Organise posts with replies for gitlab

### DISCUSSION THREAD

**Discussion by Isaac-Leonard** [magit/forge#602] [Ideas]
In Gitlab and I suspect its the same for GitHub there can be seperate conversations in an MR that have replies following each other that don't get confused with other conversations.
My issue is that it looks like forge just lists all comments / posts one by one by creation date and doesn't preserve reply order (Not talking about quote replies).
Would it be possible to change this?

**Comment by tarsius** [comment]
Actually Github issues and pull-requests are not threaded. Some types of "discussions" however are. Adding support for discussions is something I intend to do soon (~January). Once I have implemented threading for that, then I can also add it for issues/mrs on Gitlab.

**Reply by tarsius** [reply]
I have implemented this for Github discussions in summer. I haven't implemented it for Gitlab's threaded posts yet. But I still plan to do that.

(Bringing it up at this time because a duplicate was just opened (#812).)

**Comment by tarsius** [comment]
It appears this is not possible because of limitations of the REST API.

Using https://gitlab.com/testibus1/lab-test3/-/work_items/2 as an example.

First determine the `project_id` (57986043) and `iid` of that issue.

```elisp
(forge--glab-get
  (forge-get-repository "https://gitlab.com/testibus1/lab-test3")
  "/projects/:project/issues/2")
```

Get the "notes" for that issue as documented at https://docs.gitlab.com/api/notes/#list-all-issue-notes.

```elisp
(forge--glab-get
  (forge-get-repository "https://gitlab.com/testibus1/lab-test3")
  "/projects/57986043/issues/2/notes"
  )
```

I see nothing that would indicate that 3838053197 and 3838052417 are comments on 3838051762, as opposed to also directly on issue 2.

```elisp
(((id . 3838053197)
  (type . "DiscussionNote")
  (body . "comment2")
  (author ...)
  (created_at . "2026-09-15T17:40:50.059Z")
  (updated_at . "2026-09-15T17:40:50.059Z")
  (system)
  (noteable_id . 203183149)
  (noteable_type . "Issue")
  (project_id . 57986043)
  (resolvable . t)
  (resolved)
  (resolved_by)
  (resolved_at)
  (confidential)
  (internal)
  (imported)
  (imported_from . "none")
  (noteable_iid . 2)
  (commands_changes))
 ((id . 3838052417)
  (type . "DiscussionNote")
  (body . "comment")
  (author ...)
  (created_at . "2026-09-15T17:40:38.777Z")
  (updated_at . "2026-09-15T17:40:38.777Z")
  (system)
  (noteable_id . 203183149)
  (noteable_type . "Issue")
  (project_id . 57986043)
  (resolvable . t)
  (resolved)
  (resolved_by)
  (resolved_at)
  (confidential)
  (internal)
  (imported)
  (imported_from . "none")
  (noteable_iid . 2)
  (commands_changes))
 ((id . 3838051762)
  (type . "DiscussionNote")
  (body . "top")
  (author ...)
  (created_at . "2026-09-15T17:40:28.492Z")
  (updated_at . "2026-09-15T17:40:38.824Z")
  (system)
  (noteable_id . 203183149)
  (noteable_type . "Issue")
  (project_id . 57986043)
  (resolvable . t)
  (resolved)
  (resolved_by)
  (resolved_at)
  (confidential)
  (internal)
  (imported)
  (imported_from . "none")
  (noteable_iid . 2)
  (commands_changes)))
```

If someone finds another REST API endpoint that would allow me to do so, then we can reopen this.

I was planning to eventually use GraphQL for Gitlab too, but that plan is on hold. Iff I ever do that and the GraphQL API actually volunteers this information, then I would make use of that as well.
